### PR TITLE
Initial 1.4 commits

### PR DESCRIPTION
**Pull request by gargana**
*Initial Commits addressing #36 #43 #41 #39*

*Initial Vault 1.4.0 changes:*
- Support for CIS level 1 Ubuntu AMI
- Disable bash history
- Disable core dumps
- Run Vault as its own user
- Set ulimits for Vault process user
- Adding Support for Hashicorp vault 1.4 changes
- Enabling Raft storage backend
- AWS IAM Auth via IAM Role enabled on server
- Store Vault init Keys and Root token in AWS Secrets Manager

- Taskcat v9 config and CI tests
- Modernize instance types for use
- Add newer regions

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .taskcat.yml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  - ap-northeast-1
+  - sa-east-1
+  - ca-central-1
+  - ap-southeast-1
+  - ap-southeast-2
+  - eu-central-1
+  - us-east-1
+  - us-east-2
+  - us-west-2
+  parameters:
```
Comment (dcallao) [inline .taskcat.yml:25]: You should add the `AccessCIDR` param here for testing, something like `AccessCIDR: 0.0.0.0/0`? If I run `taskcat -d test run -l -e -k`, it complains: 
```
[ERROR  ] : The following parameters have no value whatsoever. The CloudFormation stack will fail to launch. Please address. str(['AccessCIDR'])
```

File: scripts/bootstrap_server.sh [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+chmod -R 0644 /etc/vault.d/*
+
+cat << EOF > /etc/environment
+export VAULT_ADDR=http://127.0.0.1:8200
+export VAULT_SKIP_VERIFY=true
+EOF
+
+. /etc/environment
+
+setcap cap_ipc_lock=+ep /usr/local/bin/vault
```
Comment (dcallao) [inline scripts/bootstrap_server.sh]: Remove this. Add the `cap_ipc_lock=+ep` in the `vault` service definition instead.

File: .gitmodules [hunk 3]
```diff
@@ -8,5 +8,9 @@
 	branch = master
 [submodule "submodules/quickstart-hashicorp-consul"]
 	path = submodules/quickstart-hashicorp-consul
-	url = https://github.com/aws-quickstart/quickstart-hashicorp-consul.git
+	url = ../../aws-quickstart/quickstart-hashicorp-consul.git
 	branch = freeze-updates
```
Comment (dcallao) [inline .gitmodules:12]: Remove `consul` git submodule - that is no longer necessary in this project

Also, In the `README` should mention the enduser should run the following commands:
```
git submodule init
git submodule update
```

File: scripts/functions.sh [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+Restart=on-failure
+PermissionsStartOnly=true
+ExecStartPre=/sbin/setcap 'cap_ipc_lock=+ep' /usr/local/bin/vault
+ExecStart=/usr/local/bin/vault server -config /etc/vault.d
+ExecReload=/bin/kill -HUP \$MAINPID
+KillSignal=SIGTERM
+User=${USER}
+Group=${GROUP}
+[Install]
+WantedBy=multi-user.target
```
Comment (dcallao) [inline scripts/functions.sh]: source: https://github.com/chrismatteson/terraform-vault-consul-deployment/blob/master/aws/install-vault.tpl
Please use this instead:
```
[Unit]
Description="HashiCorp Vault"
Documentation=https://www.vaultproject.io/docs/
Requires=network-online.target
After=network-online.target
ConditionFileNotEmpty=/etc/vault.d/vault.hcl
[Service]
User=${USER}
Group=${GROUP}
ExecStart=/opt/vault/bin/vault server -config /etc/vault.d -log-level=warn
ExecReload=/bin/kill --signal HUP $MAINPID
KillMode=process
Restart=on-failure
RestartSec=5
LimitNOFILE=65536
ProtectSystem=full
ProtectHome=read-only
PrivateTmp=yes
PrivateDevices=yes
SecureBits=keep-caps
AmbientCapabilities=CAP_IPC_LOCK
Capabilities=CAP_IPC_LOCK+ep
CapabilityBoundingSet=CAP_SYSLOG CAP_IPC_LOCK
NoNewPrivileges=yes
KillSignal=SIGINT
TimeoutStopSec=30
StartLimitInterval=60
StartLimitIntervalSec=60
StartLimitBurst=3
LimitMEMLOCK=infinity

[Install]
WantedBy=multi-user.target
```

File: scripts/bootstrap_server.sh [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+
+# Create Vault User
+user_ubuntu
+
+# Adjusting ulimits for vault user
+cat << EOF > /etc/security/limits.conf
+vault          soft    nofile          64000
+vault          hard    nofile          64000
+vault          soft    nproc           64000
+vault          hard    nproc           64000
```
Comment (dcallao) [inline scripts/bootstrap_server.sh]: Replace this table with this:
```
vault          soft    nofile          65536
vault          hard    nofile          65536
vault          soft    nproc           65536
vault          hard    nproc           65536
```

File: templates/quickstart-hashicorp-vault-master.template [hunk 6]
```diff
@@ -155,101 +182,88 @@ Parameters:
...
+  VaultClientRoleName:
+    Type: String
+    Default: client-role-iam
+    MinLength: "3"
+    ConstraintDescription: This identifier should be at least 3 characters in length.
+    Description: The internal Hashicorp Vault identifier for the AWS IAM role.
+  VaultDownloadURL:
+    Type: String
+    Description: 'The URL to download the Vault zip file'
+    Default: https://releases.hashicorp.com/vault/1.4.0/vault_1.4.0_linux_amd64.zip
```
Comment (dcallao) [inline templates/quickstart-hashicorp-vault-master.template]: We should remove this option for the enduser to insert the download URL into the wizard. We should leave the Vault version url as a static value that points to the Vault 1.4 release. This quickstart template will only run with 1.3 and above.

File: templates/quickstart-hashicorp-vault-master.template [hunk 7]
```diff
@@ -58,25 +62,44 @@ Metadata:
...
       The logical order is preserved (3 AZs are used for this deployment)'
     Type: List<AWS::EC2::AvailabilityZone::Name>
+  VaultAMIOS:
+    AllowedValues:
+      - Ubuntu-1604-HVM
+      - CIS-Ubuntu-1604-HVM
+    Default: Ubuntu-1604-HVM
+    Description: The Linux distribution for the AMI to be used for the Vault instances.
+    Type: String
   KeyPairName:
```
Comment (dcallao) [inline templates/quickstart-hashicorp-vault-master.template:103]: Insert a condition that the key must exist in order for the wizard to proceed, similar to what we had in Consul:
```
  KeyPair:
    Description: Name of an existing EC2 KeyPair to enable SSH access to the instances
    Type: "AWS::EC2::KeyPair::KeyName"
    ConstraintDescription: Must be the name of an existing EC2 KeyPair.
```

File: templates/quickstart-hashicorp-vault.template [hunk 8]
```diff
@@ -1,9 +1,101 @@
...
+    AllowedValues:
+      - Ubuntu-1604-HVM
+      - CIS-Ubuntu-1604-HVM
+    Default: Ubuntu-1604-HVM
+    Description: The Linux distribution for the AMI to be used for the Vault instances.
+    Type: String
+  KeyPairName:
     Description: Name of an existing EC2 KeyPair to enable SSH access to the instances
     Type: AWS::EC2::KeyPair::KeyName
     Default: id_rsa_aws
```
Comment (dcallao) [inline templates/quickstart-hashicorp-vault.template:102]: This must be the same as in the master template:
```
  KeyPairName:
    Description: Name of an existing EC2 KeyPair to enable SSH access to the instances
    Type: "AWS::EC2::KeyPair::KeyName"
    ConstraintDescription: Must be the name of an existing EC2 KeyPair.
```

File: templates/quickstart-hashicorp-vault.template [hunk 9]
```diff
@@ -262,23 +533,19 @@ Resources:
...
-                Fn::Sub:
-                - https://s3.amazonaws.com/${QSS3BucketName}/${QSS3KeyPrefix}scripts/cloudwatch-monitoring.sh
-                - QSS3BucketName:
-                    Ref: QSS3BucketName
-                  QSS3KeyPrefix:
-                    Ref: QSS3KeyPrefix
+                !Sub
+                  - 'https://${S3Bucket}.s3.${S3Region}.${AWS::URLSuffix}/${QSS3KeyPrefix}scripts/cloudwatch-monitoring.sh'
+                  - S3Region: !If [UsingDefaultBucket, !Ref 'AWS::Region', !Ref QSS3BucketRegion]
+                    S3Bucket: !If [UsingDefaultBucket, !Sub '${QSS3BucketName}-${AWS::Region}', !Ref QSS3BucketName]
```
Comment (dcallao) [inline templates/quickstart-hashicorp-vault.template:553]: Can you replace this S3 path with an `inline` reference to the cron config itself? `*/5 * * * * sudo /usr/local/aws-scripts-mon/mon-put-instance-data.pl --mem-util --mem-used --from-cron`

Delete file `cloudwatch-monitoring.sh` file from the `scripts` folder as well.

File: templates/quickstart-hashicorp-vault.template [hunk 10]
```diff
@@ -262,23 +533,19 @@ Resources:
...
-                Fn::Sub:
-                - https://s3.amazonaws.com/${QSS3BucketName}/${QSS3KeyPrefix}scripts/awslogs-config-file
-                - QSS3BucketName:
-                    Ref: QSS3BucketName
-                  QSS3KeyPrefix:
-                    Ref: QSS3KeyPrefix
+                !Sub
+                  - 'https://${S3Bucket}.s3.${S3Region}.${AWS::URLSuffix}/${QSS3KeyPrefix}scripts/awslogs-config-file'
+                  - S3Region: !If [UsingDefaultBucket, !Ref 'AWS::Region', !Ref QSS3BucketRegion]
+                    S3Bucket: !If [UsingDefaultBucket, !Sub '${QSS3BucketName}-${AWS::Region}', !Ref QSS3BucketName]
```
Comment (dcallao) [inline templates/quickstart-hashicorp-vault.template:563]: Remove this S3 path with an `inline` reference to the awslogs config:
```
[general]
state_file = /var/awslogs/state/agent-state

[/var/log/syslog]
file = /var/log/vault_audit.logstatus
log_group_name = __VAULT_LOG_GROUP__
log_stream_name = {instance_id}
datetime_format = %b %d %H:%M:%S
```

Delete file `awslogs-config-file`

File: templates/quickstart-hashicorp-vault.template [hunk 11]
```diff
@@ -312,364 +578,397 @@ Resources:
...
-                    Ref: QSS3BucketName
-                  QSS3KeyPrefix:
-                    Ref: QSS3KeyPrefix
+                !Sub
+                  - 'https://${S3Bucket}.s3.${S3Region}.${AWS::URLSuffix}/${QSS3KeyPrefix}scripts/vault.hcl'
+                  - S3Region: !If [UsingDefaultBucket, !Ref 'AWS::Region', !Ref QSS3BucketRegion]
+                    S3Bucket: !If [UsingDefaultBucket, !Sub '${QSS3BucketName}-${AWS::Region}', !Ref QSS3BucketName]
               mode: '000700'
               owner: root
               group: root
```
Comment (dcallao) [inline templates/quickstart-hashicorp-vault.template:599]: I thought the script `boostrap_server.sh` already configures `vault.hcl` for Vault. If not, you should remove the S3 endpoint and reference it `inline`:
```
backend "consul" {
  address = "127.0.0.1:8500"
  path    = "vault/"
}

listener "tcp" {
  address     = "0.0.0.0:8200"
  tls_disable = 1
}
```
Delete the file `vault.hcl` from the scripts folder.

File: templates/quickstart-hashicorp-vault.template [hunk 12]
```diff
@@ -312,364 +578,397 @@ Resources:
...
           commands:
             01_chmod_and_chown_vault:
-              command: chmod 0755 /usr/bin/vault && chown root:root /usr/bin/vault
+              command: chmod 0755 /usr/bin/vault && chown vault:vault /usr/bin/vault
         run_vault:
           commands:
             01_register_vault_binary:
-              command: service vault start
+              command: systemctl daemon-reload && systemctl enable vault && systemctl start vault
```
Comment (dcallao) [inline templates/quickstart-hashicorp-vault.template]: This should be 3 separate commands in order:
01_reload_daemon:
  (...)
02_enable_vault_service:
  (...)
03_start_vault_service:
  (..)
```

File: templates/quickstart-hashicorp-vault.template [hunk 13]
```diff
@@ -312,364 +578,397 @@ Resources:
...
-
-'
-            - 'easy_install https://s3.amazonaws.com/cloudformation-examples/aws-cfn-bootstrap-latest.tar.gz
+        Fn::Base64: !Sub
+          - |
+            #!/bin/bash -x
+            
+            # Install git
+            apt-get update
+            apt-get install -y git
```
Comment (dcallao) [inline templates/quickstart-hashicorp-vault.template]: Can you also install `jq` package?

File: templates/quickstart-hashicorp-vault.template [hunk 14]
```diff
@@ -312,364 +578,397 @@ Resources:
...
-          - ''
-          - - "#!/bin/bash -x\n"
-            - 'apt-get -y update
+        Fn::Base64: !Sub
+          - |
+            #!/bin/bash -x
+            
+            # Install git
+            apt-get update
+            apt-get install -y git
```
Comment (dcallao) [inline templates/quickstart-hashicorp-vault.template]: Can you also install `jq` package?

File: templates/quickstart-hashicorp-vault.template [hunk 15]
```diff
@@ -312,364 +578,397 @@ Resources:
...
+  VaultSecret:
+    Value: !Ref "VaultSecret"
+    Description: The AWS Secrets Manager Secret containing the ROOT TOKEN and Recovery Secret for Hashicorp Vault.
+  VaultClientIAMRoleArn:
+    Value: !GetAtt "VaultClientRole.Arn"
+    Description: The ARN of the AWS IAM role linked to Hashicorp Vault.
+  VaultClientIAMRoleName:
+    Value: !Ref "VaultClientRole"
+    Description: The name of the AWS IAM role linked to Hashicorp Vault.
+  # TODO: Make this configurable
```
Comment (dcallao) [inline templates/quickstart-hashicorp-vault.template]: Leave it not configurable. That's ok to post it out the outputs here.

File: templates/quickstart-hashicorp-vault.template [hunk 16]
```diff
@@ -312,364 +578,397 @@ Resources:
...
+  VaultClientRoleId:
+    Value: "client-role-iam"
```
Comment (dcallao) [inline templates/quickstart-hashicorp-vault.template]: This should be `!Ref "VaultClientRole.Id"` which points to `quickstart-client-role-iam`  - it should not be configurable.

**Review by dcallao — COMMENTED** [review]
@gargana, I have finished reviewing some minor changes. I will review the lambda stuff tomorrow morning. Thanks!

File: .taskcat.yml (hunk 1 above)
Comment (gargana) [reply .taskcat.yml:25]: I am intentionally using a .taskcat_overrides.yml to provide this and other sensitive values to my tests.
eg:
```yaml
AccessCIDR: 0.0.0.0\0
KeyPair: mykeypair
...
```
Am doing this to keep that kind of info out of the repository

Perhaps we should have a TESTING.md with information around testing with TaskCat?

File: scripts/bootstrap_server.sh (hunk 2 above)
Comment (gargana) [reply scripts/bootstrap_server.sh]: Okay noted will make this change

File: .gitmodules (hunk 3 above)
Comment (gargana) [reply .gitmodules:12]: Typically this is only for advanced users wanting to stage their own copies/modified templates from their own S3 buckets. I think perhaps we need a specific README with guidance specific to doing so. General users would launch the templates from the aws-quickstart S3 buckets and would not affect them using the templates.

File: scripts/functions.sh (hunk 4 above)
Comment (gargana) [reply scripts/functions.sh]: Will make this change

File: scripts/bootstrap_server.sh (hunk 5 above)
Comment (gargana) [reply scripts/bootstrap_server.sh]: Will make this change

File: templates/quickstart-hashicorp-vault.template (hunk 12 above)
Comment (gargana) [reply templates/quickstart-hashicorp-vault.template]: Currently not in use. Will adjust this for future use.

File: templates/quickstart-hashicorp-vault.template (hunk 13 above)
Comment (gargana) [reply templates/quickstart-hashicorp-vault.template]: I have installed JQ at an earlier point. I will try and consolidate package installations to a single point to make this less confusing.

File: templates/quickstart-hashicorp-vault.template (hunk 14 above)
Comment (gargana) [reply templates/quickstart-hashicorp-vault.template]: I have installed JQ at an earlier point. I will try and consolidate package installations to a single point to make this less confusing.

File: templates/quickstart-hashicorp-vault.template (hunk 15 above)
Comment (gargana) [reply templates/quickstart-hashicorp-vault.template]: I believe I have actually made it configurable will confirm and remove the comment.

File: templates/quickstart-hashicorp-vault.template (hunk 16 above)
Comment (gargana) [reply templates/quickstart-hashicorp-vault.template]: Can we please have a discussion around this as I am not sure I follow.

File: scripts/functions.sh (hunk 4 above)
Comment (gargana) [reply scripts/functions.sh]: This change caused problems. Vault binary at /usr/local/bin/vault adjusted this. Retesting

File: scripts/bootstrap_client.sh [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+#### For Vault Agent Auth #####
+cat << EOF > /home/ubuntu/vault-agent-wrapped.hcl
+exit_after_auth = true
+pid_file = "./pidfile"
+auto_auth {
+   method "aws" {
+       mount_path = "auth/aws"
+       config = {
+           type = "iam"
+           role = "client-role-iam"
```
Comment (dcallao) [inline scripts/bootstrap_client.sh]: The IAM role name should map out to the param that is assigned to it in the vault CFN template

File: templates/quickstart-hashicorp-vault.template (hunk 16 above)
Comment (dcallao) [reply templates/quickstart-hashicorp-vault.template]: Per our convo, leave this configurable but make this as the default name for the client role iam name: `quickstart-client-role-iam`

**Comment by dcallao** [comment]
@gargana Lambda/bash functions look good. I have noticed a few missing `TO-DO` items marked as `# TODO:`. Do you plan on completing those before the launch?

Also, as we have discussed, add these items into the the vault CFN template:
- Two yes/no options for Vault auth methods: AWS auth and Kubernetes auth. Let user choose and configure necessary params.
- Add "Take snapshot at the end of deployment? True/False" Let use choose whether True or False.

Good job! 👍

**Comment by gargana** [comment]
> * s/no options for Vault auth methods: AWS auth and Kubernetes auth. Let user choose and configure necessary params.
> * Add "Take snapshot at the end of deployment? True/False" Let use choose whether True or False.

Will be doing these in their respective issues.
Going to merge this PR.

**Pull request merged**